### PR TITLE
2788 - Introduce Vitest

### LINKED ISSUE #2788: [TESTING]: Add Jest or Vitest to `package.json`

### LINKED ISSUE #2789: [TESTING]: Configure test runner in `jest.config.js` or `vitest.config.js`

### PR DESCRIPTION
**Pull request by gcgoncalves**
# ✨ Feature / Enhancement PR

## 🔗 Epic / Issue
Closes #2788
Closes #2789 

---

## 🚀 Summary (1-2 sentences)
Introduce Vitest, configure it, and sets the base for unit testing functions on `admin.js`.

## 📋 Instructions
1. Run `npm install`
2. Run `npm run test`

### CODE REVIEW

**Comment by crivetimihai** [comment]
Thanks for bootstrapping Vitest, @gcgoncalves! The `escapeHtml` test suite is thorough — good coverage of XSS payloads and type coercion edge cases. A few items to address:

1. **MANIFEST.in regression**: The change removes `exclude eslint.config.js` and adds blanket `include *.js`, which means `eslint.config.js` and `vitest.config.js` will ship in the Python package. Please restore the eslint exclusion and add `exclude vitest.config.js`.
2. **Unused `happy-dom` dependency**: Only `jsdom` is used (in both the config and the test). `happy-dom` adds ~20 packages to the lockfile for no benefit — please remove it.
3. **CJS/ESM mixing in `setup.js`**: Line 7 uses `require("util")` while the rest uses ESM imports. Should be `import { TextEncoder, TextDecoder } from "util"`.
4. **`setup.js` vs test duplication**: The test file creates its own JSDOM instance and mocks, not using `setup.js` at all. Consider either leveraging the setup file or trimming it.

Items 1-2 should be fixed before merge; 3-4 are nice-to-haves.

File: docs/docs/test/unittest.md [hunk 1]
```diff
@@ -1,3 +1,67 @@
...
+```
+
+### Test Structure
+
+JavaScript tests are located in the `tests/` directory alongside Python tests:
+
+```
+tests/
+├── setup.js           # Vitest setup and global mocks
+├── admin.test.js      # Admin UI utility tests
```
Comment (marekdano) [inline docs/docs/test/unittest.md]: Can we have this file under a new folder `js` (or another name), something like `tests/js/admin.test.js`?

File: docs/docs/test/unittest.md [hunk 2]
```diff
@@ -1,3 +1,67 @@
...
+npm run test:ui
+```
+
+### Test Structure
+
+JavaScript tests are located in the `tests/` directory alongside Python tests:
+
+```
+tests/
+├── setup.js           # Vitest setup and global mocks
```
Comment (marekdano) [inline docs/docs/test/unittest.md]: Do we need to have `setup.js` to be under `tests` folder? 
What if we have it under the root or a new folder `js`?

File: tests/setup.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// Vitest setup file for DOM testing
+// This file runs before each test suite
+
+import { beforeEach, afterEach, vi } from "vitest";
+
+// Polyfill TextEncoder/TextDecoder for jsdom
+const { TextEncoder, TextDecoder } = require("util");
```
Comment (marekdano) [inline tests/setup.js]: Can we replace it with `import ... from` syntax?

File: package.json [hunk 4]
```diff
@@ -3,17 +3,28 @@
         "@biomejs/biome": "^2.1.1",
         "@stylistic/eslint-plugin": "^5.7.0",
         "@stylistic/stylelint-config": "^2.0.0",
+        "@vitest/coverage-v8": "^4.0.18",
+        "@vitest/ui": "^4.0.18",
         "eslint": "^9.39.2",
         "eslint-config-prettier": "^9.1.2",
         "eslint-plugin-import": "^2.32.0",
         "eslint-plugin-prettier": "^5.5.4",
+        "happy-dom": "^20.6.0",
```
Comment (marekdano) [inline package.json]: Do we need to have this package `happy-dom` here? What do we use it for?

File: vitest.config.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { defineConfig } from "vitest/config";
+
+export default defineConfig({
+    test: {
+        environment: "jsdom",
+        globals: true,
+        setupFiles: ["./tests/setup.js"],
+        include: ["tests/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}"],
```
Comment (marekdano) [inline vitest.config.js]:
```suggestion
        include: ["tests/js/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}"],
```

Then I think we won't need to have a list of test paths in the `exclude` below.

File: docs/docs/test/unittest.md (hunk 2 above)
Comment (gcgoncalves) [reply docs/docs/test/unittest.md]: I liked the idea of scoping the JS tests in a JS directory. I'll move them.

File: vitest.config.js (hunk 5 above)
Comment (gcgoncalves) [reply vitest.config.js]: Maybe after the refactor. Another option would be `tests/js/**/*.{test,spec}.{js,mjs,cjs,ts,mts,cts,jsx,tsx}`, that'll cover the `js` directory and all it's subdirectories.

File: docs/docs/test/unittest.md [hunk 6]
```diff
@@ -1,3 +1,66 @@
...
+
+### Test Structure
+
+JavaScript tests are located in the `tests/js/` directory:
+
+```
+tests/
+└── js/
+    ├── admin.test.js      # Admin UI utility tests
+    └── [future test files]
```
Comment (marekdano) [inline docs/docs/test/unittest.md]: 🚀

**Review by marekdano — APPROVED** [review]
I ran `npm run test:coverage` and I checked the coverage report at my
`file:///Users/marekdano/IBM/OSS/mcp-context-forge/coverage/lcov-report/admin.js.html` 
and in my terminal

Image

The report does not cover the tested `escapeHtml` function. 
I would expect the lines `344-356` in `admin.js` file will be covered.

This does not block merging it. We can investigate it more in our new unit tests. 

`LGTM` 🚀

**Comment by crivetimihai** [comment]
## Coverage for non-modular browser scripts in Vitest + JSDOM

### The problem

The original PR configured `@vitest/coverage-v8` with coverage reporting for `mcpgateway/static/**/*.js`, but running `vitest --coverage` reports **0% for all files**. This isn't a configuration mistake — it's a fundamental limitation of how coverage instrumentation works with non-modular browser scripts loaded via JSDOM.

### Root cause

`admin.js` is a classic browser script (not an ES module). It attaches functions to `window` and relies on browser globals (`document`, `localStorage`, `fetch`, `Chart`, etc.). The only way to test it in Node.js is to load it into JSDOM:

```javascript
const content = fs.readFileSync('admin.js', 'utf8');
dom.window.eval(content);
```

**Why every standard coverage tool fails with this pattern:**

| Tool | How it instruments | Why it can't see eval'd code |
|------|-------------------|------------------------------|
| **V8 provider** (`@vitest/coverage-v8`) | Uses `node:inspector` to collect V8 bytecode counters from the main Node.js context | JSDOM creates a separate V8 context via `vm.createContext()` — the inspector doesn't reach into sandboxed contexts |
| **Istanbul provider** (`@vitest/coverage-istanbul`) | Transforms source files during Vite's module import pipeline, injecting counter statements | `fs.readFileSync` returns a raw string — it never passes through Vite's transform pipeline, so no counters are injected |
| **c8** (process-level wrapper) | Sets `NODE_V8_COVERAGE` to capture all V8 coverage at process exit | Coverage data for eval'd/vm code has no file URL association — c8 can't map it back to source files |
| **vm.Script with filename** | Tried associating a filename via `new vm.Script(code, { filename: path })` | The script still runs in JSDOM's sandboxed context, and the coverage data stays isolated there |

All four approaches were tested and confirmed to produce 0%.

### The fix: Istanbul manual instrumentation

The solution is to instrument `admin.js` ourselves before eval'ing it, then bridge the coverage data back to Vitest:

```javascript
import { createInstrumenter } from 'istanbul-lib-instrument';

beforeAll(() => {
    // 1. Read admin.js
    const content = fs.readFileSync(adminJsPath, 'utf8');

    // 2. Instrument with Istanbul (injects __coverage__ counters)
    const instrumenter = createInstrumenter({ esModules: false });
    const instrumented = instrumenter.instrumentSync(content, adminJsPath);

    // 3. Eval instrumented code in JSDOM — counters execute in the sandbox
    dom.window.eval(instrumented);
});

afterAll(() => {
    // 4. Copy coverage data from JSDOM sandbox → Node.js globalThis
    //    where Vitest's Istanbul reporter collects it
    const jsCoverage = dom.window.__coverage__;
    if (jsCoverage) {
        globalThis['__VITEST_COVERAGE__'] = globalThis['__VITEST_COVERAGE__'] || {};
        Object.assign(globalThis['__VITEST_COVERAGE__'], jsCoverage);
    }
});
```

**Why this works:** Istanbul operates at the source level — it rewrites the JavaScript to include counter-incrementing statements *before* the code is eval'd. When the instrumented code runs inside JSDOM, the counters accumulate in `dom.window.__coverage__`. We then copy that object to `globalThis.__VITEST_COVERAGE__` (Vitest's Istanbul collection point) in `afterAll`, so it appears in the coverage report.

**Result:**
```
----------------|---------|----------|---------|---------|---
File            | % Stmts | % Branch | % Funcs | % Lines |
----------------|---------|----------|---------|---------|---
 admin.js       |    4.48 |     1.71 |    5.39 |    4.51 |
 flame-graph.js |       0 |        0 |       0 |       0 |
 gantt-chart.js |       0 |        0 |       0 |       0 |
----------------|---------|----------|---------|---------|---
```

The 4-5% coverage reflects the initialization code that runs when admin.js is eval'd plus the `escapeHtml` function exercised by the tests. As more functions get test coverage, these numbers will increase.

### Trade-offs

- **Performance**: Instrumenting 31K lines adds ~1.5s to test setup (total ~2s vs ~300ms without). Acceptable.
- **Dependencies**: Adds `istanbul-lib-instrument` and `@vitest/coverage-istanbul` as devDependencies.
- **Provider**: Must use Istanbul, not V8. The V8 provider cannot work with this approach.

### Long-term alternatives

For reference, the other approaches considered and their viability:

- **Vitest browser mode**: Run tests in a real browser via Playwright. Would give native coverage but requires browser installation in CI, and `admin.js` has heavy dependencies (Chart.js, marked, DOMPurify, htmx) that would all need to be loaded. High effort, uncertain benefit.
- **Partial ES module extraction**: Extract pure utility functions (like `escapeHtml`) into a separate `admin-utils.js` ES module that Vitest can import natively. Good long-term architecture improvement but only covers extracted functions, not the 90%+ that depends on DOM state.
- **Full ES module refactoring**: Convert all of admin.js to ES modules. Would give native coverage and better architecture, but a massive effort for 31K lines with deep DOM dependencies and inline HTML event handlers (`onclick="fn()"`).

The Istanbul manual instrumentation approach is the pragmatic solution that works today with minimal code changes.

**Comment by crivetimihai** [comment]
## Coverage Improvement Update

Added **258 new tests** across 6 categorized test files + a shared helper, bringing total from 13 → 271 tests.

### Test file breakdown

| File | Tests | Functions covered |
|------|------:|-------------------|
| `admin-validation.test.js` | 92 | `validatePassthroughHeader`, `validateInputName`, `validateUrl`, `validateJson`, `isValidIpOrCidr`, `isValidPermission`, `isValidCertificate`, `isValidBase64` |
| `admin-formatting.test.js` | 46 | `formatValue`, `formatNumber`, `formatLastUsed`, `formatFileSize`, `formatTimestamp`, `formatDate`, `truncateText` |
| `admin-parsing.test.js` | 47 | `parseUriTemplate`, `parseThinkTags`, `parseCertificateInfo`, `orderCertificateChain`, `extractTeamId`, `extractApiError`, `extractKPIData` |
| `admin-display.test.js` | 39 | `getLogLevelClass`, `getSeverityClass`, `generateStatusBadgeHtml`, `calculateSuccessRate`, `isAdminOnlyTab` |
| `admin-config.test.js` | 27 | `isAdminUser`, `getRootPath`, `getAuthenticatedUserId`, `getPerformanceAggregationConfig/Label/Query` |
| `admin-errors.test.js` | 7 | `handleFetchError` |

### Coverage numbers (admin.js)

| Metric | Before | After | Δ |
|--------|-------:|------:|---|
| Statements | 4.48% | 6.29% | +40% |
| Branches | 1.71% | 4.51% | +164% |
| Functions | 5.39% | 8.93% | +66% |
| Lines | 4.51% | 6.32% | +40% |

### Why coverage % is modest

admin.js is ~31K lines. The vast majority is DOM manipulation, `fetch` calls, Chart.js rendering, and HTMX event handlers — none of which are testable without full browser mocking or integration tests. **All testable pure/near-pure utility functions are now covered.** To push coverage further, next steps would be:

1. **DOM interaction tests** — mock `document.getElementById`, form elements, etc. for functions like `showTab`, `updateDashboard`
2. **Fetch mocking** — use `vi.fn()` to mock `window.fetch` for CRUD functions
3. **Modularize admin.js** — extract functions into ES modules so they can be imported directly without JSDOM eval

### Infrastructure improvements

- Extracted shared `tests/js/helpers/admin-env.js` helper with Istanbul instrumentation + JSDOM setup
- Added coverage counter merging so data accumulates correctly across test files
- Refactored original `admin.test.js` to use the shared helper
